canonical/main.go

```diff
@@ -0,0 +1,377 @@
+			text := scanner.Bytes()
+
+			// Trim space rather than use regex.
+			trimmed := bytes.TrimSpace(text)
+			if bytes.Equal(trimmed, startHead) {
+				inHead = true
+			}
+
+			if inHead && bytes.Contains(text, canonical) {
+				foundCanonical = true
+			}
+
+			if inHead && bytes.Equal(trimmed, endHead) {
+				if !foundCanonical {
+					link, err := canonicalLink(ctx, pkg, filename)
+					if err != nil {
+						return fmt.Errorf("canonicalLink: %v", err)
+					}
+					fmt.Fprintf(out, "    <link rel=\"canonical\" href=%q/>\n", link)
+					foundCanonical = true
+				}
+				inHead = false
+			}
+
+			out.Write(text)
+			fmt.Fprintln(out) // Text does not have a newline. Add one.
+		}
+		if err := scanner.Err(); err != nil {
+			return fmt.Errorf("error scanning %q %q: %v", name, hdr.Name, err)
+		}
+
+		// Update the header to have the right size. Otherwise, leave hdr unchanged.
+		hdr.Size = int64(out.Len())
+		tw.WriteHeader(hdr)
+		if l, err := tw.Write(out.Bytes()); err != nil || l != int(hdr.Size) {
+			return fmt.Errorf("error writing %q %q: wrote %v bytes: %v", name, hdr.Name, l, err)
+		}
+
+		atomic.AddInt64(&numFiles, 1)
+		bar.Set("prefix", fmt.Sprintf("Canonicalizing... %d files | ", numFiles))
+	}
+
+	if err := tw.Close(); err != nil {
+		return fmt.Errorf("error closing tar %v: %v", name, err)
+	}
+	if err := gw.Close(); err != nil {
+		return fmt.Errorf("error closing gzip %v: %v", name, err)
+	}
+	if err := ow.Close(); err != nil {
+		return fmt.Errorf("error closing obj %q writer: %v", name, err)
+	}
+
+	return nil
+}
+
+// parsePkg gets the package name from the tarball name.
+func parsePkg(tarballName string) string {
+	name := tarballName[len(dotnetPrefix):] // Trim docfx- prefix.
+	name = name[:strings.Index(name, "-")]  // Discard after first remaining -.
+	return name
+}
+
+// canonicalLink returns the canonical link for the given package and page.
+func canonicalLink(ctx context.Context, pkg, page string) (string, error) {
+	q := url.Values{
+		"package": []string{pkg},
+		"page":    []string{page},
+	}
+	u, err := url.Parse("https://us-central1-jonskeet-integration-tests.cloudfunctions.net/canonicalize-link")
+	if err != nil {
+		return "", err
+	}
+	u.RawQuery = q.Encode()
+	r := newRequestWithContext(ctx, http.MethodGet, u.String())
+	resp, err := httpClient.Do(r)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return string(b), nil
+}
+
+func newRequestWithContext(ctx context.Context, method, url string) *http.Request {
+	req, err := http.NewRequestWithContext(ctx, method, url, nil)
+	if err != nil {
+		panic(err)
+	}
+	return req
+}
+
+// objects returns a map of object names in the bucket.
+func objects(ctx context.Context, bucket *storage.BucketHandle) (map[string]bool, error) {
+	objs := map[string]bool{}
+	it := bucket.Objects(ctx, &storage.Query{Prefix: dotnetPrefix})
```

Comment:
If there are many objects and you're only getting the names, you can speed this up by using `ProjectionNoACL` and/or `SetAttrSelection` https://pkg.go.dev/cloud.google.com/go/storage#Query

Reply:
Done (`SetAttrSelection`).